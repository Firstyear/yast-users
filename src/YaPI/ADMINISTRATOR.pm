package YaPI::ADMINISTRATOR;

use strict;
use YaST::YCP qw(:LOGGING);
use YaPI;

textdomain ("users");

# ------------------- imported modules
YaST::YCP::Import ("MailAliases");
YaST::YCP::Import ("Users");
# -------------------------------------

our $VERSION		= '1.0.0';
our @CAPABILITIES 	= ('SLES11');
our %TYPEINFO;

=item *
C<$hash Read ();>

Returns the information about system administrator (root).
Currently return hash contains the list of mail aliases.

=cut

BEGIN{$TYPEINFO{Read} = ["function",
    [ "map", "string", "any" ]];
}
sub Read {

    my $self	= shift;

    my $root_mail	= MailAliases->GetRootAlias ();
    return {} if !defined $root_mail;

    my @root_aliases	= ();
    foreach my $alias (split (/,/, $root_mail)) {
	$alias	=~ s/[ \t]//g;
	push @root_aliases, $alias;
    }

    return {
	"aliases"	=> \@root_aliases
    }
}

=item *
C<$string Write ($argument_hash);>

write the system adminstrator data. Supported keys of the argument hash are:

    "aliases"	=> list of mail aliases
    "password"	=> new password
    
Returns error message on error.

=cut

BEGIN{$TYPEINFO{Write} = ["function",
    "string",
    [ "map", "string", "any" ]];
}
sub Write {

    my $self	= shift;
    my $args	= shift;
    my $ret	= "";

    if ($args->{"aliases"} && ref ($args->{"aliases"}) eq "ARRAY") {

	my $root_mail	= join (", ", @{$args->{"aliases"}});
	if (!MailAliases->SetRootAlias ($root_mail)) {
	    # error popup
	    $ret = __("An error occurred while setting forwarding for root's mail.");
	    return $ret;
	}
    }
    if ($args->{"password"}) {
	Users->SetRootPassword ($args->{"password"});
	Users->WriteRootPassword ();
    }
    return $ret;
}

