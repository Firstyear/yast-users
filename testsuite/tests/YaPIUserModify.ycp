/**
 * $Id$
 */

{
    // testedfiles: Users.pm UsersCache.pm USERS.pm

    include "testsuite.ycp";

    import "YaPI::USERS";
    import "Mode";

    map R = $[
	"passwd": $[
	    "local": $[
		"users": $[
		    500: $[
			"uid"		: "hh",
			"uidnumber"	: 500,
			"type"		: "local",
			"userpassword"	: "x",
			"sn"		: "HH",
			"grouplist"	: $[
			    "users"	: 1
			],
		    ],
		    "by_name": $[
			"hh":	500,
		    ],
		    "last_uid": 500,
		    "homes":	$[
			"/home/hh":	1,
		    ],
		    "usernames": $[
			"hh":		1,
		    ],
		    "uids": $[
			500:		1,
		    ]
		],
		"shadow": $[
		],
		"groups": $[
		    100: $[
			"cn"	:	"users",
			"gidnumber":	100,
			"type"	:	"local",
			"userlist"	: $[
			    "hh":	1
			],
		    ],
		    "by_name": $[
			"users":	100,
		    ],
		    "gids": $[
			100:		1,
		    ],
		    "groupnames": $[
			"users":	1,
		    ]
		],
	    ],
	    "system": $[
		"users": $[
		    0:	$[
			"uid"		: "root",
			"uidnumber"	: 0,
			"type"		: "system"
		    ],
		    "by_name": $[
			"root"		: 0
		    ],
		    "last_uid": 0,
		    "homes":	$[
		    ],
		    "usernames": $[
			"root"		: 1,
		    ],
		    "uids": $[
			0		: 1
		    ],
		],
		"shadow": $[
		],
		"groups": $[
		    0: $[
			"cn":	"root",
			"gidnumber":	0,
			"type"	:	"system"
		    ],
		    "by_name": $[
			"root"	: 0
		    ],
		    "gids": $[
			0	: 1
		    ],
		    "groupnames": $[
			"root"	: 1
		    ]
		],
	    ],
	],
	"etc" : $[
	    "fstab": [],
	    "cryptotab": [],
	    "default": $[
		"useradd": $[
		    "home":	"/tmp/local/home",
		    "groups":	"audio,video",
		    "group":	100
		]
	    ],
	],
	"target": $[
	    "stat"	: $[],
	    "size"	: 0,
	    "tmpdir"	: "/tmp/YaST"
	],
	"product": $[
		"features": $[
		    "USE_DESKTOP_SCHEDULER"	: "no",
		    "IO_SCHEDULER"		: "",
		    "ENABLE_AUTOLOGIN"		: "false",
		    "EVMS_CONFIG"		: "no"
		]
	],
    ];
    map W = $[];
    map E = $[
	"passwd" : $[
	    "init"	: true
	],
	"target": $[
	    "mkdir"	: true,
	    "bash_output"	: $[]
	]
    ];

    DUMP ("==========================================================");

    Mode::test = true;

    map<string,any> config_map  = $[
	"uid"		: "jj",
    ];
    map<string,any> data_map    = $[
	"sn"		: "HH",
    ];
    E ["target", "bash_output", "stdout"] = "jj";

    DUMP ("============ edit non-existing user 'jj': =====================");
    TEST (``(YaPI::USERS::UserModify (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ edit done ========================================");

    config_map    = $[
	"uid"           : "hh",
    ];
    E ["target", "bash_output", "stdout"] = "hh";

    DUMP ("============ edit existing user 'hh' (not really changed): ====");
    TEST (``(YaPI::USERS::UserModify (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ edit done ========================================");

    data_map["sn"]	= "HHH";

    DUMP ("============ edit existing user 'hh': (changed sn) ============");
    TEST (``(YaPI::USERS::UserModify (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ edit done ========================================");

    data_map["uid"]	= "hhh";
    E ["target", "bash_output", "stdout"] = "hhh";

    DUMP ("============ edit existing user 'hh': (changed uid) ============");
    TEST (``(YaPI::USERS::UserModify (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ edit done ========================================");

    data_map["uid"]	= "root";
    E ["target", "bash_output", "stdout"] = "root";

    DUMP ("============ edit existing user 'hh': (uid conflict) ==========");
    TEST (``(YaPI::USERS::UserModify (config_map, data_map)), [R,W,E], 0);
    DUMP ("============ edit done ========================================");

}
