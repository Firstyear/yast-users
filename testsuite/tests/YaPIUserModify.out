Dump	==========================================================
Dump	============ edit non-existing user 'jj': =====================
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.target.size "0" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/usr/lib/yp/yphelper" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/var/lib/YaST2/users.ycp" 0
Read	.target.ycp "/var/lib/YaST2/users.ycp" 0
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["grouplist":$["users":1], "sn":"HH", "type":"local", "uid":"hh", "uidnumber":500, "userpassword":"x"], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "type":"local", "userlist":$["hh":1]], "by_name":$["users":100], "gids":$[100:1], "groupnames":$["users":1]]
Read	.passwd.local.groups.by_name $["users":100]
Read	.passwd.system.users $[0:$["type":"system", "uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], "by_name":$["root":0], "gids":$[0:1], "groupnames":$["root":1]]
Read	.passwd.system.groups.by_name $["root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Return	There is no such user.
Dump	============ edit done ========================================
Dump	============ edit existing user 'hh' (not really changed): ====
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.target.size "0" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/usr/lib/yp/yphelper" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/var/lib/YaST2/users.ycp" 0
Read	.target.ycp "/var/lib/YaST2/users.ycp" 0
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["grouplist":$["users":1], "sn":"HH", "type":"local", "uid":"hh", "uidnumber":500, "userpassword":"x"], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "type":"local", "userlist":$["hh":1]], "by_name":$["users":100], "gids":$[100:1], "groupnames":$["users":1]]
Read	.passwd.local.groups.by_name $["users":100]
Read	.passwd.system.users $[0:$["type":"system", "uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], "by_name":$["root":0], "gids":$[0:1], "groupnames":$["root":1]]
Read	.passwd.system.groups.by_name $["root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash_output "echo 'hh' | grep '^[A-Za-z_][A-Za-z0-9_.-]*[A-Za-z0-9_.$-]\\?$'" $["stdout":"hh"]
Return	
Dump	============ edit done ========================================
Dump	============ edit existing user 'hh': (changed sn) ============
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.target.size "0" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/usr/lib/yp/yphelper" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/var/lib/YaST2/users.ycp" 0
Read	.target.ycp "/var/lib/YaST2/users.ycp" 0
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["grouplist":$["users":1], "sn":"HH", "type":"local", "uid":"hh", "uidnumber":500, "userpassword":"x"], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "type":"local", "userlist":$["hh":1]], "by_name":$["users":100], "gids":$[100:1], "groupnames":$["users":1]]
Read	.passwd.local.groups.by_name $["users":100]
Read	.passwd.system.users $[0:$["type":"system", "uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], "by_name":$["root":0], "gids":$[0:1], "groupnames":$["root":1]]
Read	.passwd.system.groups.by_name $["root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash_output "echo 'hh' | grep '^[A-Za-z_][A-Za-z0-9_.-]*[A-Za-z0-9_.$-]\\?$'" $["stdout":"hh"]
Execute	.target.bash "/bin/cp /etc/passwd /etc/passwd.YaST2save" 0
Write	.passwd.users $["local":$["500":$["grouplist":$["users":"1"], "modified":"edited", "org_uid":"hh", "org_uidnumber":"500", "org_user":$["grouplist":$["users":"1"], "sn":"HH", "type":"local", "uid":"hh", "uidnumber":"500", "userpassword":"x"], "plugins":[], "sn":"HHH", "type":"local", "uid":"hh", "uidnumber":"500", "userpassword":"x", "what":"edit_user"], "by_name":$["hh":"500"], "homes":$["/home/hh":"1"], "last_uid":"500", "uids":$["500":"1"], "usernames":$["hh":"1"]], "system":$["0":$["type":"system", "uid":"root", "uidnumber":"0"], "by_name":$["root":"0"], "homes":$[], "last_uid":"0", "uids":$["0":"1"], "usernames":$["root":"1"]]] true
Execute	.target.bash "/bin/cp /etc/shadow /etc/shadow.YaST2save" 0
Write	.passwd.shadow $["local":$["hh":$["userpassword":"x"]], "system":$[]] true
Execute	.target.bash "/usr/sbin/nscd -i passwd" 0
Return	
Dump	============ edit done ========================================
Dump	============ edit existing user 'hh': (changed uid) ============
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.target.size "0" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/usr/lib/yp/yphelper" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/var/lib/YaST2/users.ycp" 0
Read	.target.ycp "/var/lib/YaST2/users.ycp" 0
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["grouplist":$["users":1], "sn":"HH", "type":"local", "uid":"hh", "uidnumber":500, "userpassword":"x"], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "type":"local", "userlist":$["hh":1]], "by_name":$["users":100], "gids":$[100:1], "groupnames":$["users":1]]
Read	.passwd.local.groups.by_name $["users":100]
Read	.passwd.system.users $[0:$["type":"system", "uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], "by_name":$["root":0], "gids":$[0:1], "groupnames":$["root":1]]
Read	.passwd.system.groups.by_name $["root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash_output "echo 'hhh' | grep '^[A-Za-z_][A-Za-z0-9_.-]*[A-Za-z0-9_.$-]\\?$'" $["stdout":"hhh"]
Execute	.target.bash "/bin/cp /etc/group /etc/group.YaST2save" 0
Write	.passwd.groups $["local":$["100":$["cn":"users", "gidnumber":"100", "modified":"edited", "more_users":$["hhh":"1"], "org_cn":"users", "org_gidnumber":"100", "type":"local", "userlist":$["hhh":"1"], "what":"user_change_default"], "by_name":$["users":"100"], "gids":$["100":"1"], "groupnames":$["users":"1"]], "system":$["0":$["cn":"root", "gidnumber":"0", "type":"system"], "by_name":$["root":"0"], "gids":$["0":"1"], "groupnames":$["root":"1"]]] true
Execute	.target.bash "/bin/cp /etc/passwd /etc/passwd.YaST2save" 0
Write	.passwd.users $["local":$["500":$["grouplist":$["users":"1"], "modified":"edited", "org_uid":"hhh", "org_uidnumber":"500", "org_user":$["grouplist":$["users":"1"], "sn":"HH", "type":"local", "uid":"hh", "uidnumber":"500", "userpassword":"x"], "plugins":[], "sn":"HHH", "type":"local", "uid":"hhh", "uidnumber":"500", "userpassword":"x", "what":"edit_user"], "by_name":$["hh":"500"], "homes":$["/home/hh":"1"], "last_uid":"500", "uids":$["500":"1"], "usernames":$["hh":"1"]], "system":$["0":$["type":"system", "uid":"root", "uidnumber":"0"], "by_name":$["root":"0"], "homes":$[], "last_uid":"0", "uids":$["0":"1"], "usernames":$["root":"1"]]] true
Execute	.target.bash "/bin/cp /etc/shadow /etc/shadow.YaST2save" 0
Write	.passwd.shadow $["local":$["hhh":$["userpassword":"x"]], "system":$[]] true
Execute	.target.bash "/usr/sbin/nscd -i passwd" 0
Execute	.target.bash "/usr/sbin/nscd -i group" 0
Return	
Dump	============ edit done ========================================
Dump	============ edit existing user 'hh': (uid conflict) ==========
Read	.target.tmpdir "/tmp/YaST"
Read	.etc.default.useradd."expire" 0
Read	.etc.default.useradd."group" 100
Read	.etc.default.useradd."groups" "audio,video"
Read	.etc.default.useradd."home" "/tmp/local/home"
Read	.etc.default.useradd."inactive" 0
Read	.etc.default.useradd."shell" 0
Read	.etc.default.useradd."skel" 0
Read	.target.string "/etc/shells" 0
Read	.target.size "0" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/usr/lib/yp/yphelper" 0
Read	.etc.nsswitch_conf.passwd 0
Read	.target.size "/var/lib/YaST2/users.ycp" 0
Read	.target.ycp "/var/lib/YaST2/users.ycp" 0
Execute	.passwd.init $["base_directory":"/etc", "max_system_gid":"499", "max_system_uid":"499"] true
Read	.passwd.local.users $[500:$["grouplist":$["users":1], "sn":"HH", "type":"local", "uid":"hh", "uidnumber":500, "userpassword":"x"], "by_name":$["hh":500], "homes":$["/home/hh":1], "last_uid":500, "uids":$[500:1], "usernames":$["hh":1]]
Read	.passwd.local.users.by_name $["hh":500]
Read	.passwd.local.shadow $[]
Read	.passwd.local.groups $[100:$["cn":"users", "gidnumber":100, "type":"local", "userlist":$["hh":1]], "by_name":$["users":100], "gids":$[100:1], "groupnames":$["users":1]]
Read	.passwd.local.groups.by_name $["users":100]
Read	.passwd.system.users $[0:$["type":"system", "uid":"root", "uidnumber":0], "by_name":$["root":0], "homes":$[], "last_uid":0, "uids":$[0:1], "usernames":$["root":1]]
Read	.passwd.system.users.by_name $["root":0]
Read	.passwd.system.shadow $[]
Read	.passwd.system.groups $[0:$["cn":"root", "gidnumber":0, "type":"system"], "by_name":$["root":0], "gids":$[0:1], "groupnames":$["root":1]]
Read	.passwd.system.groups.by_name $["root":0]
Read	.passwd.passwd.pluslines 0
Read	.passwd.shadow.pluslines 0
Read	.passwd.group.pluslines 0
Read	.passwd.local.users.last_uid 500
Read	.passwd.local.users.homes $["/home/hh":1]
Read	.passwd.local.users.usernames $["hh":1]
Read	.passwd.local.users.uids $[500:1]
Read	.passwd.system.users.last_uid 0
Read	.passwd.system.users.homes $[]
Read	.passwd.system.users.usernames $["root":1]
Read	.passwd.system.users.uids $[0:1]
Read	.passwd.local.groups.gids $[100:1]
Read	.passwd.local.groups.groupnames $["users":1]
Read	.passwd.system.groups.gids $[0:1]
Read	.passwd.system.groups.groupnames $["root":1]
Execute	.target.bash_output "echo 'root' | grep '^[A-Za-z_][A-Za-z0-9_.-]*[A-Za-z0-9_.$-]\\?$'" $["stdout":"root"]
Return	There is a conflict between the entered
Dump	============ edit done ========================================
